### PR TITLE
BAEL-6323 : Count Number of Rows in JPA

### PR DESCRIPTION
**Pull request by imsurajmishra**
JIRA: https://jira.baeldung.com/browse/BAEL-6323
Draft: https://drafts.baeldung.com/?p=164139&preview=true

### CODE REVIEW

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import org.junit.jupiter.api.Test;
+
+import com.baeldung.countrows.AccountStatsApplication;
+import com.baeldung.countrows.entity.Account;
+import com.baeldung.countrows.entity.Permission;
+import com.baeldung.countrows.repository.AccountRepository;
+import com.baeldung.countrows.repository.PermissionRepository;
+import com.baeldung.countrows.service.AccountStatsLogic;
+
+import org.springframework.beans.factory.annotation.Autowired;
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:22]: Please, make sure you are using the [baeldung formatted](https://github.com/Baeldung/dev-settings/blob/main/javaCodeStyle/intellij/intellij-baeldung-style.xml)

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    private PermissionRepository permissionRepository;
+
+    @Autowired
+    private AccountRepository accountRepository;
+
+    @Autowired
+    private AccountStatsLogic accountStatsLogic;
+
+    @AfterEach
+    public void afterEach(){
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:38]: add space between () {

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        accountRepository.deleteAll();
+        permissionRepository.deleteAll();
+    }
+
+    @Test
+    public void givenAccountInTable_whenPerformCount_returnsAppropriateCount(){
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:44]: add space between () {

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    }
+    @Test
+    public void givenAccountInTable_whenPerformCountByPermissionAndCreatedOnUsingJPQL_returnsAppropriateCount() throws ParseException {
+        savePermissions();
+        Account account = saveAccount();
+        long count = accountStatsLogic.getAccountsByPermissionAndCreatedOnUsingJPQL(account.getPermission(), account.getCreatedOn());
+        assertThat(count).isEqualTo(1);
+    }
+
+    private Account saveAccount(){
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:112]: add space between () {

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        return accountRepository.save(getAccount());
+    }
+
+    private  void savePermissions(){
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:116]: add space between () {

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/service/AccountStatsLogic.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    private EntityManager entityManager;
+
+    @Autowired
+    private PermissionRepository permissionRepository;
+
+    public long getAccountCount(){
+        return accountRepository.count();
+    }
+
+    public long getAccountCountByUsername(String username){
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/service/AccountStatsLogic.java:36]: add space between () {

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        savePermissions();
+        Account account = saveAccount();
+        long count = accountStatsLogic.getAccountCountByPermissionAndCreatedOn(account.getPermission(), account.getCreatedOn());
+        assertThat(count).isEqualTo(1);
+    }
+
+    @Test
+    public void givenAccountInTable_whenPerformCountUsingCQ_returnsAppropriateCount() throws ParseException {
+        savePermissions();
+        Account account = saveAccount();
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:69]: If you are not using the **Account account**, let's remove it

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        long count = accountStatsLogic.getAccountsByPermissionAndCreateOnUsingCQ(account.getPermission(), account.getCreatedOn());
+        assertThat(count).isEqualTo(1);
+    }
+
+    @Test
+    public void givenAccountInTable_whenPerformCountUsingJPQL_returnsAppropriateCount() throws ParseException {
+        savePermissions();
+        Account account = saveAccount();
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:93]: If you are not using the **Account account**, let's remove it

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        long count = accountStatsLogic.getAccountsUsingJPQL();
+        assertThat(count).isEqualTo(1);
+    }
+    @Test
+    public void givenAccountInTable_whenPerformCountByPermissionUsingJPQL_returnsAppropriateCount() throws ParseException {
+        savePermissions();
+        Account account = saveAccount();
+        long count = accountStatsLogic.getAccountsByPermissionUsingJPQL(account.getPermission());
+        assertThat(count).isEqualTo(1);
+    }
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:103]: add break line

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
+        Date parsedDate = dateFormat.parse("2023-04-29");
+        return parsedDate;
+    }
+
+    private Account getAccount() {
+        Permission permission = permissionRepository.findByType("admin");
+        Account account = new Account();
+        String seed = UUID.randomUUID().toString();
+        account.setUsername("username_"+seed);
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:136]: add space between +

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+import javax.persistence.*;
+
+import java.security.PrivateKey;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.util.Date;
+
+@Entity
+@Table(name="ACCOUNTS")
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java:11]: add spaces between =

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+    @OneToOne
+    @JoinColumn(name = "permissions_id")
+    private Permission permission;
+
+    public int getUserId() {
+        return userId;
+    }
+
+    public void setUserId(int userId) {
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java:32]: Not used anywhere, we can remove it.

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+    public String getUsername() {
+        return username;
+    }
+
+    public void setUsername(String username) {
+        this.username = username;
+    }
+
+    public String getPassword() {
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java:44]: Not used anywhere, we can remove it.

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+    public Timestamp getCreatedOn() {
+        return createdOn;
+    }
+
+    public void setCreatedOn(Timestamp createdOn) {
+        this.createdOn = createdOn;
+    }
+
+    public Timestamp getLastLogin() {
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java:68]: Not used anywhere, we can remove it.

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
+package com.baeldung.countrows.entity;
+
+import javax.persistence.*;
+
+@Entity
+@Table(name="PERMISSIONS")
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java:6]: add spaces between =

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+public class Permission {
+
+    @Id
+    @GeneratedValue(strategy= GenerationType.SEQUENCE, generator = "permissions_id_sq")
+    @SequenceGenerator(name = "permissions_id_sq", sequenceName = "permissions_id_sq", allocationSize = 1)
+    private int id;
+
+    private String type;
+
+    public int getId() {
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java:16]: Not used anywhere, we can remove it.

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+        return id;
+    }
+
+    public void setId(int id) {
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java:20]: Not used anywhere, we can remove it.

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+        this.id = id;
+    }
+
+    public String getType() {
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java:24]: Not used anywhere, we can remove it.

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import javax.persistence.*;
+
+@Entity
+@Table(name="PERMISSIONS")
+public class Permission {
+
+    @Id
+    @GeneratedValue(strategy= GenerationType.SEQUENCE, generator = "permissions_id_sq")
+    @SequenceGenerator(name = "permissions_id_sq", sequenceName = "permissions_id_sq", allocationSize = 1)
+    private int id;
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java:12]: Not used anywhere, we can remove it.

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/service/AccountStatsLogic.java [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    public long getAccountsByPermissionAndCreatedOnUsingJPQL(Permission permission, Date date) throws ParseException {
+        Query query = entityManager.createQuery("SELECT COUNT(*) FROM Account a WHERE a.permission = ?1 and a.createdOn > ?2");
+        query.setParameter(1, permission);
+        query.setParameter(2, new Timestamp(date.getTime()));
+        return (long) query.getSingleResult();
+    }
+
+    private static Date getDate() throws ParseException {
```
Comment (vinipx) [inline persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/service/AccountStatsLogic.java:116]: Not used anywhere, we can remove it.

**Review by vinipx — CHANGES_REQUESTED** [review]
Please, make sure you are using the [baeldung formatted](https://github.com/Baeldung/dev-settings/blob/main/javaCodeStyle/intellij/intellij-baeldung-style.xml)

After importing it, you can navigate your files and do **ctrl + l + o + i** to adjust all.

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.Date;
+
+@Entity
+@Table(name="ACCOUNTS")
+public class Account {
+    @Id
+    @GeneratedValue(strategy= GenerationType.SEQUENCE, generator = "accounts_seq")
+    @SequenceGenerator(name = "accounts_seq", sequenceName = "accounts_seq", allocationSize = 1)
+    @Column(name = "user_id")
+    private int userId;
```
[the remark this answers is not in the document]
Comment (imsurajmishra) [reply persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Account.java:17]: this is primary_key and my understanding is that  we cannot remove it

File: persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java (hunk 19 above)
Comment (imsurajmishra) [reply persistence-modules/spring-boot-persistence-3/src/main/java/com/baeldung/countrows/entity/Permission.java:12]: this is primary_key and my understanding is that we cannot remove it

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+        assertThat(count).isEqualTo(1);
+    }
+
+    @Test
+    public void givenAccountInTable_whenPerformCountUsingJPQL_returnsAppropriateCount() throws ParseException {
+        savePermissions();
+        Account account = saveAccount();
+        long count = accountStatsLogic.getAccountsUsingJPQL();
+        assertThat(count).isEqualTo(1);
+    }
```
[the remark this answers is not in the document]
Comment (imsurajmishra) [reply persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:96]: added.

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java (hunk 9 above)
Comment (imsurajmishra) [reply persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:104]: added

File: persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java (hunk 1 above)
Comment (imsurajmishra) [reply persistence-modules/spring-boot-persistence-3/src/test/java/com/baeldung/boot/countrows/accountstatslogic/AccountStatsUnitTest.java:22]: done

**Review by vinipx — APPROVED** [review]

**Pull request merged**